ocs/cluster.py

```diff
@@ -0,0 +1,248 @@
+        # Dynamically added attribute 'port'
+        pod.port = l2
+        logging.info(f"port={pod.port}")
+        return pod
+
+    def mon_change_count(self, new_count):
+        """
+        Change mon count in the cluster
+
+        Args:
+            new_count(int): Absolute number of mons required
+        """
+        self.cluster.data['spec']['mon']['count'] = new_count
+        self.cluster.apply(**self.cluster.data)
+        self.mon_health_check(new_count)
+        logger.info(f"Mon count changed to {new_count}")
+        self.mon_count = new_count
+        self.scan_cluster()
+        self.cluster.reload()
+
+    def mon_health_check(self, count):
+        """
+        TODO: Mon pod count alone can't tell monitor health
+            need other parameters as well
+        """
+        try:
+            assert POD.wait_for_resource(
+                condition='Running', selector=self.mon_selector,
+                resource_count=count,
+            )
+        except AssertionError as e:
+            logger.error(e)
+            raise MonCountException(
+                f"Failed to achieve desired Mon count"
+                f" {count}"
+            )
+
+    def mds_change_count(self, new_count):
+        """
+        Change mds count in the cluster
+
+        Args:
+            new_count(int): Absolute number of mds required
+        """
+        self.cephfs.data['spec']['metadataServer']['activeCount'] = new_count
+        logger.info(self.cephfs.data)
+        self.cephfs.apply(**self.cephfs.data)
+        self.mds_health_check(new_count)
+        logger.info(f"MDS count changed to {new_count}")
+        self.mds_count = new_count
+        self.scan_cluster()
+        self.cephfs.reload()
+
+    def mds_health_check(self, count):
+        """
+        TODO: mds count alone can't tell whether fs is healthy
+                need some more parameters as well
+        """
+        try:
+            assert POD.wait_for_resource(
+                condition='Running', selector=self.mds_selector,
+                resource_count=count
+        )
+        except AssertionError as e:
+            logger.error(e)
+            raise MDSCountException(
+                f"Failed to achieve desired MDS count"
+                f" {count}"
+            )
+
+    def get_admin_key(self):
+        """
+        Returns:
+            adminkey(str): base64 encoded key
+        """
+        return self.get_user_key('client.admin')
+
+    def get_user_key(self, user):
+        """
+        Args:
+            user(str): ceph username ex: client.user1
+        Returns:
+            key(str): base64 encoded user key
+        """
+        out = self.toolbox[0].exec_cmd_on_pod(
+            f"ceph auth get-key {user} --format json"
+        )
+        if 'ENOENT' in out:
+            return False
+        key_base64 = base64.b64encode(out['key'].encode()).decode()
+        return key_base64
+
+    def create_user(self, username, caps):
```

Comment:
why do we need to create things directly in ceph?

Reply:
Usually in plain ceph world if you consider an example of CephFs there is a feature for subdirectory exports for specific users, something like how home directory works in linux. So A user can see only his home directory and that starts as root for him . To achieve this we need a specific user say client.user1 who has permission to access only /user1 directory from Cephfs and client.user2 who can see and access only /user2 . I hope this will appear as a supported feature in OCS as well. Look at this chunk of yaml from cephfs.yaml

```yaml
  # For provisionVolume: "true":
  #   A new volume will be created along with a new Ceph user.
  #   Requires admin credentials (adminID, adminKey).
  # For provisionVolume: "false":
  #   It is assumed the volume already exists and the user is expected
  #   to provide path to that volume (rootPath) and user credentials (userID, userKey).
  provisionVolume: "true"

  # Ceph pool into which the volume shall be created
  # Required for provisionVolume: "true"
  pool: cephfs_data

  # Root path of an existing CephFS volume
  # Required for provisionVolume: "false"
  # rootPath: /absolute/path
```

In the above snippet if ```ProvisionVolume:False``` it expects us to already create an user along with key. I just thought just extra 4 lines can be useful in this scenario hence added it. I am not sure if secret.yamls can create users for us .

Reply:
Also I don't see a point in always using client.admin for all the operations. In real world definitely non-admin(ceph) users will be consuming the storage.